sdk/core/core/src/az_curl_adapter.c

```diff
@@ -10,28 +10,248 @@
+  }
+  char * buffer = (char *)p_writable_buffer;
+
+  // write buffer
+  az_span const writable_buffer
+      = (az_span const){ .begin = p_writable_buffer, .size = buffer_size };
+  az_result const write_result = az_write_to_buffer(writable_buffer, p_header, separator);
+
+  // attach header only when write was OK
+  if (az_succeeded(write_result)) {
+    p_headers->p_list = curl_slist_append(p_headers->p_list, buffer);
+  }
+  // at any case, error or OK, free the allocated memory
+  free(p_writable_buffer);
+  return write_result;
+}
+
+/**
+ * @brief loop all the headers from a HTTP request and set each header into easy curl
+ *
+ * @param p_hrb
+ * @param p_headers
+ * @return az_result
+ */
+az_result az_build_headers(
+    az_http_request_builder * const p_hrb,
+    az_curl_headers_list * p_headers) {
 
-AZ_NODISCARD az_result az_headers_to_curl(az_headers_data * const p_state, az_pair const header) {
-  const az_span_seq token_seq = az_build_header_callback(&header);
-  char * str_header;
-  AZ_RETURN_IF_FAILED(az_span_seq_to_new_str(token_seq, &str_header));
+  az_pair header;
+  for (uint16_t offset = 0; offset < p_hrb->headers_end; ++offset) {
+    AZ_RETURN_IF_FAILED(az_http_request_builder_get_header(p_hrb, offset, &header));
+    AZ_RETURN_IF_FAILED(az_add_header_to_curl_list(&header, p_headers, AZ_HEADER_SEPARATOR_STR));
+  }
 
-  p_state->p_list = curl_slist_append(p_state->p_list, str_header);
+  return AZ_OK;
+}
 
-  free(str_header);
+/**
+ * @brief writes a url request adds a cero to make it a c-string. Return error if any of the write
+ * operations fails.
+ *
+ * @param writable_buffer
+ * @param url_from_request
+ * @return az_result
+ */
+az_result az_write_url(az_span const writable_buffer, az_const_span const url_from_request) {
+  az_span_builder writer = az_span_builder_create(writable_buffer);
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, url_from_request));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, AZ_STR("\0")));
   return AZ_OK;
 }
 
-AZ_NODISCARD az_result
-az_build_headers(az_http_request const * const p_request, az_headers_data * p_headers) {
-  // create callback for visitor
-  az_pair_visitor const pair_visitor = az_headers_to_curl_callback(p_headers);
+/**
+ * @brief This is the function that curl will use to write response into a user provider span
+ * Function receives the size of the response and must return this same number, otherwise it is
+ * consider that function failed
+ *
+ * @param contents
+ * @param size
+ * @param nmemb
+ * @param userp
+ * @return int
+ */
+int write_to_span(void * contents, size_t size, size_t nmemb, void * userp) {
+  size_t const realsize = size * nmemb + 1;
+  az_span * const user_buffer = (az_span * const)userp;
+
+  // handle error when response won't feat user buffer
+  if (user_buffer->size < realsize) {
+    fprintf(stderr, "response size is greater than user buffer for writing response");
+    return AZ_ERROR_HTTP_FAILED_REQUEST;
```

Comment:
This is a callback that gets invoked by curl. Are you sure you can give AZ_ERROR to curl?